sklearn/gaussian_process/gpc.py

```diff
@@ -0,0 +1,180 @@
+
+        return f_star > 0
+
+    def predict_proba(self, X):
+        # Based on Algorithm 3.2 of GPML
+        K_star = \
+            self.kernel.cross(self.X_fit_, X)  # K_star =k(x_star)
+        f_star = K_star.T.dot(self.y_fit_ - self.pi)  # Line 4
+        v = solve(self.L, self.W_sr.dot(K_star))  # Line 5
+        var_f_star = self.kernel.auto(X) - v.T.dot(v)  # Line 6
+
+        # Line 7:
+        # Approximate \int log(z) * N(z | f_star, var_f_star)
+        # Approximation is due to Williams & Barber, "Bayesian Classification
+        # with Gaussian Processes", Appendix A:
+        # Approximate the logistic sigmoid by a linear combination of
+        # 5 error functions
+        # See
+        # blitiri.blogspot.de/2012/11/gaussian-integral-of-error-function.html
+        # for information on how this integral can be computed
+        alpha = 1 / (2 * np.diag(var_f_star))
+        gamma = self.lambdas * f_star
+        integrals = np.sqrt(np.pi / alpha) \
+            * erf(gamma * np.sqrt(alpha / (alpha + self.lambdas**2))) \
+            / (2 * np.sqrt(np.diag(var_f_star) * 2 * np.pi))
+        pi_star = (self.coefs * integrals).sum(axis=0) + .5 * self.coefs.sum()
+
+        return pi_star
+
+    def log_marginal_likelihood(self, theta, eval_gradient=False):
+        import copy  # XXX
+        kernel = copy.deepcopy(self.kernel)
+        kernel.params = theta
+
+        if eval_gradient:
+            K, K_gradient = \
+                kernel.auto(self.X_fit_, eval_gradient=True)
+        else:
+            K = kernel.auto(self.X_fit_)
+
+        K[np.diag_indices_from(K)] += self.jitter
+
+        # Return temporaries
+        f, Z, (pi, W_sr, L, b, a) = \
+            self._posterior_mode(K, return_temporaries=True)
+
+        if not eval_gradient:
+            return Z
+
+        # Compute gradient based on Algorithm 5.1 of GPML
+        d_Z = np.empty(theta.shape[0])
+        R = W_sr.dot(cho_solve((L, True), W_sr))  # Line 7
+        C = solve(L, W_sr.dot(K))  # Line 8
+        # Line 9:
+        s_2 = -0.5*(np.diag(K) - np.diag(C.T.dot(C))) \
+            * (pi * (1 - pi) * (1 - 2*pi))  # third derivative
+        for j in range(d_Z.shape[0]):
+            C = K_gradient[..., j]   # Line 11
+            s_1 = .5 * a.T.dot(C).dot(a) - .5 * np.trace(R.dot(C))  # Line 12
+
+            b = C.dot(self.y_fit_ - pi)  # Line 13
+            s_3 = b - K.dot(R).dot(b)  # Line 14
+
+            d_Z[j] = s_1 + s_2.T.dot(s_3)  # Line 15
+
+        return Z, d_Z
+
+    def _posterior_mode(self, K, return_temporaries=False):
+        # Based on Algorithm 3.1 of GPML
+        f = np.zeros_like(self.y_fit_, dtype=np.float64)
+        log_marginal_likelihood = -np.inf
+        while True:
+            # Line 4
+            pi = 1 / (1 + np.exp(-f))
+            W = -np.diag(-pi*(1-pi))  # XXX: avoid creating square matrix?
+            # Line 5
+            W_sr = np.sqrt(W)
+            B = np.eye(W.shape[0]) + W_sr.dot(K).dot(W_sr)
```

Comment:
You could also write, if `pi` is 1D and `W_sr = np.sqrt(pi * (1-pi))`, the product `W_sr[:, np.newaxis] * K * W_sr` which would be faster to evaluate. In any case, you should take the `sqrt` before the `diag` operation if you wish to stick with `diag`